blockchain/message_pool/src/msgpool.rs

```diff
@@ -776,11 +877,303 @@ where
+    for actor in local_addrs.iter() {
+        if let Some(mset) = pending.read().await.get(actor) {
+            if mset.msgs.is_empty() {
+                continue;
+            }
+            let mut pend: HashMap<u64, SignedMessage> = HashMap::with_capacity(mset.msgs.len());
+            for (nonce, m) in mset.msgs.clone().into_iter() {
+                pend.insert(nonce, m);
+            }
+            pending_map.insert(*actor, pend);
+        }
+    }
+    drop(local_addrs);
+
+    if pending_map.is_empty() {
+        return Ok(());
+    }
+
+    let mut chains = Vec::new();
+    for (actor, mset) in pending_map.iter() {
+        let mut next = create_message_chains(&api, actor, mset, &base_fee_lower_bound, &ts).await?;
+        chains.append(&mut next);
+    }
+
+    if chains.is_empty() {
+        return Ok(());
+    }
+
+    chains.sort_by(|a, b| match a.before(b) {
+        true => Ordering::Greater,
+        false => Ordering::Less,
+    });
+
+    let mut msgs: Vec<SignedMessage> = vec![];
+    let mut gas_limit = types::BLOCK_GAS_LIMIT;
+    let mut i = 0;
+    'l: while i < chains.len() {
+        let msg_chain = &mut chains[i];
+        let chain = msg_chain.curr().unwrap().clone();
+        if msgs.len() > REPUB_MSG_LIMIT {
+            break;
+        }
+        // TODO: Check min gas
+
+        // check if chain has been invalidated
+        if !chain.valid {
+            i += 1;
+            continue;
+        }
+
+        // check if fits in block
+        if chain.gas_limit <= gas_limit {
+            // check the baseFee lower bound -- only republish messages that can be included in the chain
+            // within the next 20 blocks.
+            for m in chain.msgs.iter() {
+                if !allow_negative_chains(ts.epoch()) && m.gas_fee_cap() < &base_fee_lower_bound {
+                    msg_chain.invalidate();
+                    continue 'l;
+                }
+                gas_limit -= m.gas_limit();
+                msgs.push(m.clone());
+            }
+            i += 1;
+            continue;
+        }
+        msg_chain.trim(gas_limit, &base_fee, true);
+        let mut j = i;
+        while j < chains.len() - 1 {
+            if chains[j].before(&chains[j + 1]) {
+                break;
+            }
+            chains.swap(j, j + 1);
+            j += 1;
+        }
+    }
+    drop(ts);
+    let mut count = 0;
+    for m in msgs.iter() {
+        let mb = m.marshal_cbor()?;
+        network_sender
+            .send(NetworkMessage::PubsubMessage {
+                topic: PUBSUB_MSG_TOPIC.clone(),
+                message: mb,
+            })
+            .await;
+        count += 1;
+        if count < msgs.len() {
+            // this delay is here to encourage the pubsub subsystem to process the messages serially
+            // and avoid creating nonce gaps because of concurrent validation
+            // TODO: I am not sure if we will run into the same issues
```

Comment:
uh, I don't think we will, all requests are passed in through the channel and the ordering is sync compatible. They only do this because their pubsub publishing is accessible anywhere and memory unsafe I think. (100 ms sleep on every message on every republish scares me)